### ISSUE TITLE
FEValues cleanup: improve data locality in get_data()/fill_fe_values()

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#1821]
The various finite element classes have lots of code of the following form:
```
      for (unsigned int k=0; k<n_q_points; ++k)
        {
          poly_space.compute(quadrature.point(k),
                             values, grads, grad_grads,
                             third_derivatives,
                             fourth_derivatives);

          if (update_flags & update_values)
            {
              if (inverse_node_matrix.n_cols() == 0)
                for (unsigned int i=0; i<this->dofs_per_cell; ++i)
                  data->shape_values[i][k] = values[i];
              else
                for (unsigned int i=0; i<this->dofs_per_cell; ++i)
                  {
                    Tensor<1,dim> add_values;
                    for (unsigned int j=0; j<this->dofs_per_cell; ++j)
                      add_values += inverse_node_matrix(j,i) * values[j];
                    data->shape_values[i][k] = add_values;
                  }
            }

          if (update_flags & update_gradients)
            {
              if (inverse_node_matrix.n_cols() == 0)
                for (unsigned int i=0; i<this->dofs_per_cell; ++i)
                  data->shape_grads[i][k] = grads[i];
              else
```
From a data locality point of view, this is inefficient since there are all of these if-statements in the overall loop over `k`, and we access different data objects one after the other. It would be more efficient to pull the loop over `k` under the various `if (update_flags|...)` statements and duplicate the loop header.

I'm working on other things in the same area right now, so don't want to do it right away, but will get to it when that is done.

Related to #1198.

**Comment by bangerth** [comment]
#1845 addresses this to the most part. There is a piece in fe_system.cc I'd like to fix but that conflicts with the currently open #1841.

**Comment by bangerth** [comment]
Fixed by #1845.
